faunadb/client.go

```diff
@@ -127,20 +142,58 @@ func (client *FaunaClient) prepareRequest(expr Expr) (request *http.Request, err
 		if request, err = http.NewRequest("POST", client.endpoint, bytes.NewReader(body)); err == nil {
 			request.Header.Add("Authorization", client.basicAuth)
 			request.Header.Add("Content-Type", "application/json; charset=utf-8")
+			client.addLastTxnTimeHeader(request)
+		}
+	}
+
+	return
+}
+
+func (client *FaunaClient) parseResponse(response *http.Response) (value Value, err error) {
+	var parsedResponse Value
+
+	if err = client.storeLastTxnTime(response.Header); err == nil {
+		if parsedResponse, err = parseJSON(response.Body); err == nil {
+			value, err = parsedResponse.At(resource).GetValue()
 		}
 	}
 
 	return
 }
 
-func (client *FaunaClient) parseResponse(response *http.Response) (Value, error) {
-	value, err := parseJSON(response.Body)
+func (client *FaunaClient) addLastTxnTimeHeader(request *http.Request) {
+	if client.isTxnTimeEnabled {
+		if lastSeen := atomic.LoadInt64(&client.lastTxnTime); lastSeen != 0 {
+			request.Header.Add("X-Last-Seen-Txn", strconv.FormatInt(lastSeen, 10))
+		}
+	}
+}
 
-	if err != nil {
-		return nil, err
+func (client *FaunaClient) storeLastTxnTime(header http.Header) (err error) {
+	if client.isTxnTimeEnabled {
+		var newTxnTime int64
+		var present bool
+
+		if newTxnTime, present, err = parseTxnTimeHeader(header); present && err == nil {
+			for {
+				oldTxnTime := atomic.LoadInt64(&client.lastTxnTime)
+				if oldTxnTime > newTxnTime ||
+					atomic.CompareAndSwapInt64(&client.lastTxnTime, oldTxnTime, newTxnTime) {
+					break
+				}
+			}
+		}
 	}
 
-	return value.At(resource).GetValue()
+	return
+}
+
+func parseTxnTimeHeader(header http.Header) (txnTime int64, present bool, err error) {
+	if lastSeenHeader := header.Get("X-Txn-Time"); lastSeenHeader != "" {
+		txnTime, err = strconv.ParseInt(lastSeenHeader, 10, 64)
+		present = true
```

Comment:
think its safe to just assume that the default `int64` value (0) is sufficient to determine whether its present? We'll never see a valid `X-Txn-Time` as 0 AFAIK.

Reply:
actually we kinda do that test anyways when setting the header so should we just drop the `present` flag?